### ISSUE TITLE
Use of `heapq` in liota/core/metric_handler.py

### ISSUE THREAD

**Issue by bfrggit** [vmware/liota#140]
On Ln 58 `def put_and_notify(self, item, block=True, timeout=None)`, it seems that it gets the smallest item every time using `heapq.nsmallest(1, self.queue)[0]` (Ln 64,87). It is confusing that we are using a "PriorityQueue" (Ln 52) but still needs `heapq.nsmallest` to sort them.

**Comment by bfrggit** [comment]
No one seems to have looked at this. However, while I am profiling a Liota script with dummy sampling function, I found RegisteredMetric.__cmp__() and __lt__() (one of the six functions that Python 3 provided to replace __cmp__) calls took most of running time. I am doing more profiling now, but I suspect it has something to do with this potential misuse.

PriorityQueue should be able to maintain its contents in order by just doing put() and get(), this line looks like we are sorting everything again even when they are already in place.

I am trying to do some assertion based test to see if this line returns exactly the same thing with calling get() directly. However I am having a little multi-threading concern here since even if they are completely equivalent in result, calling on after the other could cause someone else to change the queue in between and end up with different results... Let's see then...

**Comment by bfrggit** [comment]
Adding some profiling results (top time consuming function calls). Experiment was creating 500 dummy metrics collecting at 1 Hz.

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   117457    0.499    0.000  202.748    0.002 heapq.py:461(nsmallest)
 58528996   51.282    0.000  169.471    0.000 registered_metric.py:140(RegisteredMetric.__lt__)
 58528996   76.294    0.000  118.189    0.000 registered_metric.py:133(RegisteredMetric.__cmp__)
        1    0.035    0.035   44.673   44.673 multi_metric.py:34(<module>)
      503    0.003    0.000   42.911    0.085 datetime.py:1030(__new__)
      501    0.007    0.000   42.907    0.086 __init__.py:196(c_double)
      502    0.045    0.000   42.893    0.085 __init__.py:192(c_float)
      504    0.251    0.000   42.815    0.085 tempfile.py:555(_TemporaryFileCloser)
      502    0.010    0.000   41.866    0.083 __init__.py:249(c_bool)
      504    0.676    0.001   41.694    0.083 __init__.py:244(c_void_p)
      508    6.073    0.012   40.128    0.079 __init__.py:401(suspend_hooks)
  6434620    4.523    0.000   31.063    0.000 encoder.py:411(_iterencode)
6434620/19291212   14.828    0.000   26.540    0.000 encoder.py:332(_iterencode_dict)
 58278498   23.142    0.000   23.142    0.000 misc.py:23(cmp)
  6426440    6.771    0.000   17.478    0.000 encoder.py:276(_iterencode_list)
48949/48950    0.759    0.000   12.069    0.000 client.py:934(Client.loop)
    38816    0.177    0.000    8.111    0.000 registered_metric.py:122(RegisteredMetric.send_data)
    38816    0.224    0.000    7.768    0.000 dcc.py:80(IotControlCenter.publish)
    12561    0.131    0.000    5.423    0.000 client.py:1257(Client.loop_read)
116450/116480    0.997    0.000    5.416    0.000 queue.py:147(Queue.get)
    39322    0.119    0.000    5.302    0.000 mqtt_dcc_comms.py:156(MqttDccComms.send)
    51355    0.744    0.000    5.282    0.000 client.py:1757(Client._packet_read)
    39322    0.289    0.000    5.183    0.000 mqtt.py:298(Mqtt.publish)
```

**Comment by bfrggit** [comment]
Did a very brief test with the same dummy script above.

There are four similar lines like Ln 64,87 in `metric_handler.py`, for each of them, I added
```python
debug_first = self.queue[0]
assert(first_element == debug_first) # What ever the original variable is called
```
Ran my script with 500 metrics for 5 minutes, no AssertionError was received in log.

I was wrong at the beginning... get() does not meet the need because get() will remove the first element instead of getting it. However, self.queue[0] does the trick for a heap based PriorityQueue, because the 0-th element in the heap is always the smallest one.

Reference link: https://stackoverflow.com/questions/9287919/can-i-get-an-item-from-a-priorityqueue-without-removing-it-yet

**Warning just in case:** self.queue[1] will not guarantee to return the item ranked second, not for any other index than 0.
